Include "computed QoS" instead of "status QoS" in validation error when in-place resource updates result in recalculated QoS

jkyros opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Currently: 
- In-place resource updates can result in Pod QoS changing due to recalculation, which causes the update to fail validation
- That validation failure error currently includes the existing QoS value from the Pod's status, rather than the new computed QoS

This PR: 
- changes the error message to output the computed QoS, so the user can know what the QoS value was recalculated to and hopefully correlate it to their attempted resource changes 

Example: 

Take this pod: 
```
apiVersion: v1
kind: Pod
metadata:
  name: sleeper
  namespace: default
spec:
  containers:
  - args:
    - -c
    - sleep infinity
    command:
    - /bin/sh
    image: registry.k8s.io/ubuntu-slim:0.1
    imagePullPolicy: IfNotPresent
    name: sleeper
    resources:
      requests:
        cpu: 25m
        memory: 100Mi
```
Remove the resources to make it recalculate to `BestEffort`: 
```
$ kubectl patch pod sleeper --type="strategic" -p'{"spec":{"containers":[{"name":"sleeper","resources":{"$patch":"delete"}}]}}'
```
Observe the validation output (the pod is already `Burstable`, but the error says `Burstable` is the invalid value): 
```
The Pod "sleeper" is invalid: 
* metadata: Invalid value: "Burstable": Pod QoS is immutable
* spec: Forbidden: pod updates may not change fields other than `spec.containers[*].image`,`spec.initContainers[*].image`,`spec.activeDeadlineSeconds`,`spec.tolerations` (only additions to existing tolerations),`spec.terminationGracePeriodSeconds` (allow it to be set to 1 if it was previously negative)
```

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
In-place resource updates that recalculate pod QoS now properly include the recalculated QoS value with the validation failure
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

jkyros committed:
Log computed QoS on error instead of status QoS

Pod QoS is immutable, but it is a computed field, so in-place resource
changes can result in updates being rejected because they would change
the QoS class -- e.g. removing all resources from a Burstable pod would
re-compute its QoS class as BestEffort, so it's not allowed.

When we return an error for a re-computed QoS, we erroneously log the
existing/status QoS class rather than the new computed one, which makes
it very confusing, e.g. a Burstable pod getting re-calculated as
BestEffort would log that Qos could not be changed to Burstable because
it was immutable.

This commit just changes the output such that it properly logs the
computed QoS that couldn't be applied, which should make it easier for
the user to understand why their resource changes are impossible.

Signed-off-by: John Kyros <jkyros@redhat.com>

jkyros commented:
/sig node

bart0sh commented:
/triage accepted
/priority important-longterm
/ok-to-test
/cc

bart0sh commented:
/retest

iholder101 commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5136,8 +5136,9 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
 
-	if qos.GetPodQOS(oldPod) != qos.ComputePodQOS(newPod) {
-		allErrs = append(allErrs, field.Invalid(fldPath, newPod.Status.QOSClass, "Pod QoS is immutable"))
+	computedNewQOS := qos.ComputePodQOS(newPod)
+	if qos.GetPodQOS(oldPod) != computedNewQOS {
+		allErrs = append(allErrs, field.Invalid(fldPath, computedNewQOS, "Pod QoS is immutable"))
```
nit
```suggestion
	if computedNewQOS := qos.ComputePodQOS(newPod); qos.GetPodQOS(oldPod) != computedNewQOS {
		allErrs = append(allErrs, field.Invalid(fldPath, computedNewQOS, "Pod QoS is immutable"))
```

iholder101 commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5136,8 +5136,9 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
 
-	if qos.GetPodQOS(oldPod) != qos.ComputePodQOS(newPod) {
-		allErrs = append(allErrs, field.Invalid(fldPath, newPod.Status.QOSClass, "Pod QoS is immutable"))
+	computedNewQOS := qos.ComputePodQOS(newPod)
+	if qos.GetPodQOS(oldPod) != computedNewQOS {
+		allErrs = append(allErrs, field.Invalid(fldPath, computedNewQOS, "Pod QoS is immutable"))
```
Is it possible/easy to understand if there's an in-place scaling taking place? If it is, the problem can be made much clearer if we mention this was caused due to a resize IMO. WDYT?

jkyros replied:
Yes, I think the only time this message can come out is when there is an in-place scaling taking place because:
-  the function we're in ( `ValidatePodUpdate`)  seems to only validate the spec, while the QoS is in the status 
-  we're comparing "old" QoS to "computed new" QoS, so "new"  QoS is ignored

So I think at least here it's just kind of quietly ignoring the case where you directly change the QoS. 

So hmm, keeping with the tone of the other validation errors, maybe something like `pod resource changes may not change pod QoS, pod QoS is immutable` ?

hshiina commented:
There is a test case that tries to change a QoS class at resizing in e2e tests. Modifying this string might be helpful to verify this fix:
https://github.com/kubernetes/kubernetes/blob/6e8e1f53b09328849f655d38eae9bacbbeb3445e/test/e2e/node/pod_resize.go#L1448

jkyros commented:
Thanks @hshiina, I've aligned this to match the given feedback:  
- Changed message wording
- Updated the message in the matching `pod_resize.go` test 
- Test case was not enforcing message, it looked like it was just in there for output,  so I made it enforce that the message matched -- I can change it back if we don't think we want to do that, I just figured it would help someone know to update the test of we change it again 

The new output looks like: 
```
The Pod "sleeper" is invalid: 
* metadata: Invalid value: "BestEffort": pod resource changes may not change computed pod QoS, pod QoS is immutable
* spec: Forbidden: pod updates may not change fields other than `spec.containers[*].image`,`spec.initContainers[*].image`,`spec.activeDeadlineSeconds`,`spec.tolerations` (only additions to existing tolerations),`spec.terminationGracePeriodSeconds` (allow it to be set to 1 if it was previously negative)
```

jkyros commented:
there were apparently also some unit tests that I missed, updated those too  :smile:

iholder101 commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5130,8 +5130,8 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
-	if qos.GetPodQOS(oldPod) != qos.ComputePodQOS(newPod) {
-		allErrs = append(allErrs, field.Invalid(fldPath, newPod.Status.QOSClass, "Pod QoS is immutable"))
+	if computedNewQOS := qos.ComputePodQOS(newPod); qos.GetPodQOS(oldPod) != computedNewQOS {
+		allErrs = append(allErrs, field.Invalid(fldPath, computedNewQOS, "pod resource changes may not change computed pod QoS, pod QoS is immutable"))
```
Perhaps it would be helpful to also show both the old and new QoS classes?

```suggestion
		allErrs = append(allErrs, field.Invalid(fldPath, computedNewQOS, fmt.Sprintf("pod QoS was changed from %s to %s. Resource changes may not change computed pod QoS, pod QoS is immutable", qos.GetPodQOS(oldPod), computedNewQOS)))
```

bart0sh replied:
+1
@jkyros please apply the suggestion, thanks!

iholder101 reviewed (commented):
Sorry for the late reply.
Looks great to me! Left a minor suggestion, but it's not a must.

/lgtm
